src/cdk/explorer/detectCdkProjects.ts

```diff
@@ -0,0 +1,98 @@
+import { readDirAsString } from '../../shared/filesystemUtilities'
+
+export interface CdkProject {
+    workspaceFolder: vscode.WorkspaceFolder
+    cdkJsonPath: string
+    treePath: string
+}
+
+export async function detectCdkProjects(
+    workspaceFolders: vscode.WorkspaceFolder[] | undefined = vscode.workspace.workspaceFolders
+): Promise<CdkProject[]> {
+    if (!workspaceFolders) {
+        return []
+    }
+
+    return (await Promise.all(workspaceFolders.map(detectCdkProjectsFromWorkspaceFolder))).reduce(
+        (accumulator: CdkProject[], current: CdkProject[]) => {
+            accumulator.push(...current)
+
+            return accumulator
+        },
+        []
+    )
+}
+
+async function detectCdkProjectsFromWorkspaceFolder(workspaceFolder: vscode.WorkspaceFolder): Promise<CdkProject[]> {
+    const result = []
+
+    for await (const treeJson of detectLocalCdkProjects({ workspaceUris: [workspaceFolder.uri] })) {
+        const cdkJsonPath = path.join(treeJson.fsPath, '../', 'cdk.json')
+        const project = { workspaceFolder: workspaceFolder, cdkJsonPath: cdkJsonPath, treePath: treeJson.fsPath }
+        result.push(project)
+    }
+
+    return result
+}
+
+export interface DetectCdkProjectsContext {
+    access: typeof access
+    readDir: typeof readDirAsString
+    stat: typeof stat
+}
+
+class DefaultDetectCdkProjectsContext implements DetectCdkProjectsContext {
+    public readonly access = access
+    public readonly readDir = readDirAsString
+    public readonly stat = stat
+}
+
+export async function* detectLocalCdkProjects({
+    workspaceUris,
+    context = new DefaultDetectCdkProjectsContext()
+}: {
+    workspaceUris: vscode.Uri[]
+    context?: DetectCdkProjectsContext
+}): AsyncIterableIterator<vscode.Uri> {
+    for (const workspaceFolder of workspaceUris) {
+        for await (const folder of getFolderCandidates(context, workspaceFolder)) {
+            yield* detectCdkProjectsInFolder(context, folder)
+        }
+    }
+}
+
+async function* getFolderCandidates(context: DetectCdkProjectsContext, uri: vscode.Uri): AsyncIterableIterator<string> {
+    // Search the root and first level of children only.
+    yield uri.fsPath
+
+    const entries = await context.readDir(uri.fsPath)
+    for (const entry of entries.map(p => path.join(uri.fsPath, p))) {
+        const stats = await context.stat(entry)
+        if (stats.isDirectory()) {
+            yield entry
+        }
+    }
+}
+
+async function* detectCdkProjectsInFolder(
+    context: DetectCdkProjectsContext,
+    folder: string
+): AsyncIterableIterator<vscode.Uri> {
+    for (const treeJsonPath of [path.join(folder, 'cdk.out/tree.json')]) {
+        try {
+            await context.access(treeJsonPath)
+            yield vscode.Uri.file(treeJsonPath)
+        } catch (err) {
+            // This is usually because the file doesn't exist, but could also be a permissions issue.
+            // TODO: Log at most verbose (i.e. 'silly') logging level.
```

Comment:
```
getLogger().debug(`Error detecting CDK Projects in ${folder}`, err)
```

Reply:
^ Also, initialize the logger at the start of the file (`const logger = getLogger()`) and use that throughout the file

Reply:
Since this is a top level function (and not within a class), getLogger should be called within the function, not at the top of the file. Whether it is called in the catch block or at the start of the function is up to you.